cmd/fake-lepton/main.go

```diff
@@ -126,23 +129,41 @@ func connectToSocket(dbusService *service) error {
 }
 
 func sendCPTV(conn *net.UnixConn, file string) error {
-    mu.Lock()
-    defer mu.Unlock()
-    fullpath := path.Join(cptvDir, file)
-    if _, err := os.Stat(fullpath); err != nil {
-        fmt.Printf("%v does not exist\n", fullpath)
-        return err
-    }
-    log.Printf("sending raw frames of %v\n", fullpath)
-    r, err := cptv.NewFileReader(fullpath)
-    if err != nil {
-        return err
+    select {
+    case lockChannel <- struct{}{}:
+        fullpath := path.Join(cptvDir, file)
+        if _, err := os.Stat(fullpath); err != nil {
+            log.Printf("%v does not exist\n", fullpath)
+            unlock()
```

Comment:
Can you not just defer unlock?   That would be much less error prone.  And defer the unlock int his function not in the one you call.

Reply:
Because sendCPTV is a go routine (asynchronous) it would unlock straight away, so it does have to be done this way